`optional any` generates uncompilable C++ and Rust code

### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/fory/issues) and found no similar issues.


### Version

Latest Fory

### Component(s)

Rust, C++

### Minimal reproduce step

Generate C++ and Rust code using this IDL:

```
package issue.optional_any;

message Holder {
  optional any value = 1;
}
```
and compile the generated code with g++ and cargo.

Note: since Fory C++ compiler only generate header files for IDL, we should define and compile the simple source file like this:

```c++
#include "issue_optional_any.h"

int main() {
  issue::optional_any::Holder holder;
  return holder.has_value() ? 1 : 0;
}
```

### What did you expect to see?

Code should be compiled.

### What did you see instead?

Compilation failed for both C++ and Rust.

g++ compilation error:

```text
In file included from /tmp/fory-cpp-optional-any.3A0fEE/compile.cc:1:
/tmp/fory-cpp-optional-any.3A0fEE/issue_optional_any.h: In member function ‘const std::any& issue::optional_any::Holder::value() const’:
/tmp/fory-cpp-optional-any.3A0fEE/issue_optional_any.h:52:12: error: no match for ‘operator*’ (operand type is ‘const std::any’)
   52 |     return *value_;
      |            ^~~~~~~
```

cargo compilation error:

```text
error[E0277]: the trait bound `dyn Any + Send + Sync: Default` is not satisfied
    --> src/issue_optional_any.rs:24:5
     |
  21 | #[derive(::fory::ForyStruct, Default)]
     |                              ------- in this derive macro expansion
...
  24 |     pub value: ::std::sync::Arc<dyn ::std::any::Any + Send + Sync>,
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Default` is not implemented for `dyn Any + Send + Sync`
     |
help: the following other types implement trait `Default`
    --> /home/hpy/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/sync.rs:3715:1
     |
3715 | impl<T: Default> Default for Arc<T> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Arc<T>`
...
3767 | impl Default for Arc<str> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^ `Arc<str>`
...
3782 | impl Default for Arc<core::ffi::CStr> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Arc<CStr>`
...
3801 | impl<T> Default for Arc<[T]> {
     | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Arc<[T]>`
     = note: required for `Arc<dyn Any + Send + Sync>` to implement `Default`

error[E0599]: no method named `is_some` found for struct `Arc<(dyn Any + Send + Sync + 'static)>` in the current scope
  --> src/issue_optional_any.rs:31:23
   |
31 |         if self.value.is_some() {
   |                       ^^^^^^^ method not found in `Arc<(dyn Any + Send + Sync + 'static)>`

Some errors have detailed explanations: E0277, E0599.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `optional_any_repro` (lib) due to 2 previous errors
```



### Anything Else?

## Root cause

Let's analyze the generated code.

The generated C++ code:

```c++
class Holder final {
  public:

  bool has_value() const {
    return value_.has_value();
  }

  const std::any& value() const {
    return *value_;
  }

  void set_value(std::any value) {
    value_ = std::move(value);
  }

  void clear_value() {
    value_.reset();
  }

  bool operator==(const Holder& other) const {
    return ((!value_.has_value() && !other.value_.has_value()) || (value_.type() == other.value_.type()));
  }

  private:
    std::any value_;

  public:
  FORY_STRUCT(Holder, (value_, fory::F(1).nullable()));
};
```

We can see Fory generates C++ `std::any` for IDL `optional any`. This is correct behavior as defined in the doc:

https://github.com/apache/fory/blob/4660351b625da3222f7c4665454cdec2c552f419/docs/compiler/schema-idl.md?plain=1#L1391

But the real issue here is that `generate_field_accessors()` chooses the optional-field accessor path from `field.optional`, and emits

```cpp
return *value_;
```
https://github.com/apache/fory/blob/4660351b625da3222f7c4665454cdec2c552f419/compiler/fory_compiler/generators/cpp.py#L811-L812

This expression is valid for `std::optional<T>`, but invalid for `std::any` since it is not dereferenceable.

And the generated Rust code:

```rust
#[derive(::fory::ForyStruct, Default)]
pub struct Holder {
    #[fory(id = 1, nullable = true)]
    pub value: ::std::sync::Arc<dyn ::std::any::Any + Send + Sync>,
}

impl ::std::fmt::Debug for Holder {
    fn fmt(&self, f: &mut ::std::fmt::Formatter<'_>) -> ::std::fmt::Result {
        f.write_str("Holder { ")?;
        f.write_str("value: ")?;
        if self.value.is_some() {
            f.write_str("any(...)")?;
        } else {
            f.write_str("None")?;
        }
        f.write_str(" }")
    }
}
```

In Rust, `optional any` still generates a bare `Arc<dyn Any + Send + Sync>` field instead of an `Option<...>`, which is the same as C++.

But other Rust generation paths still look only at `field.optional`:

- `Default` is derived because optional fields are assumed to be defaultable.
  https://github.com/apache/fory/blob/4660351b625da3222f7c4665454cdec2c552f419/compiler/fory_compiler/generators/rust.py#L1060-L1061

- The custom `Debug` impl emits `self.value.is_some()` because optional fields are
  assumed to be `Option<T>`.
  https://github.com/apache/fory/blob/4660351b625da3222f7c4665454cdec2c552f419/compiler/fory_compiler/generators/rust.py#L1258-L1259

`Option` has the `is_some()` method and implements `Debug`, but not for `Arc<dyn Any + Send + Sync>`.


## Suggested fix

At first I though of fixing each edge cases in both Rust and C++ compiler case by case.
But I didn't know whether `optional any` would trigger edge cases in other languages.

The doc already says `any` as writing a null flag, same as a
nullable field. Therefore, `optional any` appears to be redundant with plain
`any`. 

So instead of adding language-specific edge-case handling for every generated path,
why won't we canonicalize `optional any` directly into `any` in the IR?

e.g. 

```
optional any value = 1;
```

should be normalized to:

```
any value = 1;
```

This would fix above edge cases directly.


### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/compiler/fory_compiler/ir/validator.py
+++ b/compiler/fory_compiler/ir/validator.py
@@ -47,6 +47,7 @@ INVALID_MAP_KEY_KINDS = {
     PrimitiveKind.DECIMAL,
 }
 INVALID_MAP_KEY_MESSAGE = "map keys do not support any, binary, float, decimal, message, union, list, map, or array types"
+OPTIONAL_ANY_MESSAGE = "optional or nullable any is not supported; use any instead"
 
 
 @dataclass
@@ -81,6 +82,7 @@ class SchemaValidator:
         self._check_type_references()
         self._check_services()
         self._check_collection_type_rules()
+        self._check_optional_any_rules()
         if not self.allow_nested_collections:
             self._check_collection_nesting()
         self._check_ref_rules()
@@ -627,6 +629,40 @@ class SchemaValidator:
         for union in self.schema.unions:
             for f in union.fields:
                 check_type(f.field_type, f)
+
+    def _check_optional_any_rules(self) -> None:
+        def is_any_type(field_type: FieldType) -> bool:
+            return (
+                isinstance(field_type, PrimitiveType)
+                and field_type.kind == PrimitiveKind.ANY
+            )
+
+        def check_field(field: Field) -> None:
+            field_type = field.field_type
+            if field.optional and is_any_type(field_type):
+                self._error(OPTIONAL_ANY_MESSAGE, field.location)
+
+            if isinstance(field_type, ListType):
+                if field_type.element_optional and is_any_type(field_type.element_type):
+                    self._error(OPTIONAL_ANY_MESSAGE, field.location)
+            elif isinstance(field_type, MapType):
+                if field_type.value_optional and is_any_type(field_type.value_type):
+                    self._error(OPTIONAL_ANY_MESSAGE, field.location)
+
+        def check_message_fields(message: Message) -> None:
+            for f in message.fields:
+                check_field(f)
+            for nested_msg in message.nested_messages:
+                check_message_fields(nested_msg)
+            for nested_union in message.nested_unions:
+                for f in nested_union.fields:
+                    check_field(f)
+
+        for message in self.schema.messages:
+            check_message_fields(message)
+        for union in self.schema.unions:
+            for f in union.fields:
+                check_field(f)
 
     def _check_collection_nesting(self) -> None:
         def check_field(

--- a/compiler/fory_compiler/tests/test_xlang_type_system.py
+++ b/compiler/fory_compiler/tests/test_xlang_type_system.py
@@ -24,7 +24,11 @@ from fory_compiler.frontend.proto import ProtoFrontend
 from fory_compiler.ir.ast import ArrayType, ListType, MapType, PrimitiveType
 from fory_compiler.ir.emitter import FDLEmitter
 from fory_compiler.ir.types import PrimitiveKind
-from fory_compiler.ir.validator import SchemaValidator
+from fory_compiler.ir.validator import (
+    INVALID_MAP_KEY_MESSAGE,
+    OPTIONAL_ANY_MESSAGE,
+    SchemaValidator,
+)
 
 
 def parse_schema(source: str):
@@ -188,11 +192,7 @@ def test_map_rejects_non_portable_key_types(key_type):
     )
 
     assert not ok
-    assert any(
-        "map keys do not support any, binary, float, decimal, message, union, list, map, or array types"
-        in err.message
-        for err in validator.errors
-    )
+    assert any(INVALID_MAP_KEY_MESSAGE in err.message for err in validator.errors)
 
 
 @pytest.mark.parametrize(
@@ -271,11 +271,7 @@ def test_map_rejects_message_and_union_key_types(source):
     _schema, validator, ok = validate_schema(source)
 
     assert not ok
-    assert any(
-        "map keys do not support any, binary, float, decimal, message, union, list, map, or array types"
-        in err.message
-        for err in validator.errors
-    )
+    assert any(INVALID_MAP_KEY_MESSAGE in err.message for err in validator.errors)
 
 
 @pytest.mark.parametrize(
@@ -316,6 +312,59 @@ def test_map_rejects_message_and_union_key_types(source):
     ],
 )
 def test_map_accepts_enum_key_types(source):
+    _schema, validator, ok = validate_schema(source)
+
+    assert ok, validator.errors
+
+
+@pytest.mark.parametrize(
+    "source",
+    [
+        """
+        message Invalid {
+            optional any value = 1;
+        }
+        """,
+        """
+        message Invalid {
+            any value = 1 [nullable = true];
+        }
+        """,
+        """
+        message Invalid {
+            list<optional any> values = 1;
+        }
+        """,
+        """
+        message Invalid {
+            map<string, optional any> values = 1;
+        }
+        """,
+    ],
+)
+def test_optional_any_is_rejected(source):
+    _schema, validator, ok = validate_schema(source)
+
+    assert not ok
+    assert any(OPTIONAL_ANY_MESSAGE in err.message for err in validator.errors)
+
+
+@pytest.mark.parametrize(
+    "source",
+    [
+        """
+        message Valid {
+            any value = 1;
+        }
+        """,
+        """
+        message Valid {
+            optional list<any> values = 1;
+        }
+        """,
+    ],
+)
+def test_non_direct_optional_any_is_accepted(source):
     _schema, validator, ok = validate_schema(source)
 
     assert ok, validator.errors

--- a/docs/compiler/schema-idl.md
+++ b/docs/compiler/schema-idl.md
@@ -993,9 +993,12 @@ list_type    := 'list' '<' { 'optional' | 'ref' | scalar_encoding } field_type '
 array_type   := 'array' '<' array_element_type '>'
 ```
 
-`optional` before `list` applies to the collection field. `ref` is only valid
-for named message/union fields; for collection contents, use `list<ref T>` or
-`map<K, ref V>`. `repeated` is accepted as an alias for `list`.
+`optional` before `list` applies to the collection field. `optional` is not
+supported when it is applied directly to `any`; use `any`, `list<any>`, or
+`map<K, any>` instead of `optional any`, `list<optional any>`, or
+`map<K, optional any>`. `ref` is only valid for named message/union fields; for
+collection contents, use `list<ref T>` or `map<K, ref V>`. `repeated` is
+accepted as an alias for `list`.
 
 ### Field Modifiers
 
@@ -1009,6 +1012,10 @@ message User {
     optional string email = 2; // Nullable
 }
 ```
+
+Do not use `optional` or `[nullable = true]` directly on `any`. The compiler
+rejects `optional any`, `any [nullable = true]`, `list<optional any>`, and
+`map<K, optional any>`; use `any`, `list<any>`, or `map<K, any>` instead.
 
 **Generated Code:**
 
@@ -1389,6 +1396,9 @@ message Envelope [id=122] {
 **Notes:**
 
 - `any` always writes a null flag (same as `nullable`) because values may be empty.
+- `optional` and `[nullable = true]` are not supported directly on `any`; use
+  `any`, `list<any>`, or `map<K, any>` instead of `optional any`,
+  `list<optional any>`, or `map<K, optional any>`.
 - Allowed dynamic values are limited to `bool`, `string`, `enum`, `message`, and `union`.
   Other primitives (numeric, bytes, date/time) and list/map are not supported; wrap them in a
   message or use explicit fields instead.

